vowpalwabbit/core/src/reductions/eigen_memory_tree.cc

```diff
@@ -0,0 +1,815 @@
+  {
+    float score = scorer_predict(b, base, ex, *example);
+
+    if (score < best_score)
+    {
+      best_score = score;
+      best_example = example.get();
+    }
+  }
+
+  return best_example;
+}
+
+void node_predict(emt_tree& b, single_learner& base, emt_node& cn, emt_example& ex, VW::example& ec)
+{
+  auto* closest_ex = node_pick(b, base, cn, ex);
+  ec.pred.multiclass = (closest_ex != nullptr) ? closest_ex->label : 0;
+  ec.loss = (ec.l.multi.label != ec.pred.multiclass) ? ec.weight : 0;
+}
+
+void emt_predict(emt_tree& b, single_learner& base, VW::example& ec)
+{
+  b.all->ignore_some_linear = false;
+  emt_example ex(*b.all, &ec);
+
+  emt_node& cn = *tree_route(b, ex);
+  node_predict(b, base, cn, ex, ec);
+  tree_bound(b, &ex);
+}
+
+void emt_learn(emt_tree& b, single_learner& base, VW::example& ec)
+{
+  b.all->ignore_some_linear = false;
+  auto ex = VW::make_unique<emt_example>(*b.all, &ec);
+
+  emt_node& cn = *tree_route(b, *ex);
+  scorer_learn(b, base, cn, *ex, ec.weight);
+  node_predict(b, base, cn, *ex, ec);  // vw learners predict and emt_learn
+
+  tree_bound(b, ex.get());
+  node_insert(cn, std::move(ex));
+  node_split(b, cn);
+}
+
+#ifdef VW_ENABLE_EMT_DEBUG_TIMER
+void emt_end_pass_timer(emt_tree& b)
+{
+  *(b.all->trace_message) << "##### pass_time: " << static_cast<float>(clock() - b.begin) / CLOCKS_PER_SEC << std::endl;
+
+  b.begin = clock();
+}
+#endif
+
+}  // namespace
+/////////////////end of eigen_memory_tree///////////////////
+////////////////////////////////////////////////////////////
+
+///////////////////Save & Load//////////////////////////////////////
+////////////////////////////////////////////////////////////////////
+namespace VW
+{
+namespace model_utils
+{
+
+size_t read_model_field(io_buf& io, reductions::eigen_memory_tree::emt_example& ex)
+{
+  size_t bytes = 0;
+  bytes += read_model_field(io, ex.base);
+  bytes += read_model_field(io, ex.full);
+  bytes += read_model_field(io, ex.label);
+  return bytes;
+}
+size_t write_model_field(
+    io_buf& io, const reductions::eigen_memory_tree::emt_example& ex, const std::string& upstream_name, bool text)
+{
+  size_t bytes = 0;
+  bytes += write_model_field(io, ex.base, upstream_name + "base", text);
+  bytes += write_model_field(io, ex.full, upstream_name + "full", text);
+  bytes += write_model_field(io, ex.label, upstream_name + "label", text);
+  return bytes;
+}
+
+size_t read_model_field(io_buf& io, reductions::eigen_memory_tree::emt_node& node)
+{
+  size_t bytes = 0;
+  bytes += read_model_field(io, node.router_decision);
+  bytes += read_model_field(io, node.left);
+  bytes += read_model_field(io, node.right);
+  bytes += read_model_field(io, node.router_weights);
+  bytes += read_model_field(io, node.examples);
+  return bytes;
+}
+
+size_t write_model_field(
+    io_buf& io, const reductions::eigen_memory_tree::emt_node& node, const std::string& upstream_name, bool text)
+{
+  size_t bytes = 0;
```

Comment:
readable model looks like that now:
```
...
emtrootrightrightrightright.is_null() = false
emtrootrightrightrightrightrouter_decision = 0.3114859461784363
emtrootrightrightrightrightleft.is_null() = false
emtrootrightrightrightrightleftrouter_decision = 0
emtrootrightrightrightrightleftleft.is_null() = true
emtrootrightrightrightrightleftright.is_null() = true
emtrootrightrightrightrightleftrouter_weights.size() = 0
emtrootrightrightrightrightleftexamples.size() = 4
emtrootrightrightrightrightleftexamples[0].is_null() = false
emtrootrightrightrightrightleftexamples[0]base.size() = 3
emtrootrightrightrightrightleftexamples[0]base[0].first = 2
emtrootrightrightrightrightleftexamples[0]base[0].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]base[1].first = 100
emtrootrightrightrightrightleftexamples[0]base[1].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]base[2].first = 116060
emtrootrightrightrightrightleftexamples[0]base[2].second = 1
emtrootrightrightrightrightleftexamples[0]full.size() = 3
emtrootrightrightrightrightleftexamples[0]full[0].first = 2
emtrootrightrightrightrightleftexamples[0]full[0].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]full[1].first = 100
emtrootrightrightrightrightleftexamples[0]full[1].second = 0.07031249
emtrootrightrightrightrightleftexamples[0]full[2].first = 116060
emtrootrightrightrightrightleftexamples[0]full[2].second = 1
emtrootrightrightrightrightleftexamples[0]label = 1
emtrootrightrightrightrightleftexamples[1].is_null() = false
emtrootrightrightrightrightleftexamples[1]base.size() = 3
emtrootrightrightrightrightleftexamples[1]base[0].first = 2
emtrootrightrightrightrightleftexamples[1]base[0].second = 0.062499993
emtrootrightrightrightrightleftexamples[1]base[1].first = 100
emtrootrightrightrightrightleftexamples[1]base[1].second = 0.062499993
emtrootrightrightrightrightleftexamples[1]base[2].first = 116060
emtrootrightrightrightrightleftexamples[1]base[2].second = 1
emtrootrightrightrightrightleftexamples[1]full.size() = 3
emtrootrightrightrightrightleftexamples[1]full[0].first = 2
...
```
should we separate upstream from field name with '.'?

Reply:
Yeah I think that would help readability a lot here